src/group_theory/subgroup/basic.lean

```diff
@@ -2584,6 +2584,48 @@ end subgroup
 
 end actions
 
+/-! ### Mul-opposite subgroups -/
+
+section mul_opposite
+
+namespace subgroup
+
+/-- A subgroup `H` of `G` determines a subgroup `H.opposite` of the opposite group `Gᵐᵒᵖ`. -/
+@[to_additive "An additive subgroup `H` of `G` determines an additive subgroup `H.opposite` of the
+  opposite additive group `Gᵃᵒᵖ`."]
+def opposite (H : subgroup G) : subgroup Gᵐᵒᵖ :=
+{ carrier := mul_opposite.unop ⁻¹' (H : set G),
+  one_mem' := H.one_mem,
+  mul_mem' := λ a b ha hb, H.mul_mem hb ha,
+  inv_mem' := λ a, H.inv_mem }
+
+/-- Bijection between a subgroup `H` and its opposite. -/
+@[to_additive "Bijection between an additive subgroup `H` and its opposite.", simps]
+def opposite_equiv (H : subgroup G) : H ≃ H.opposite :=
+mul_opposite.op_equiv.subtype_equiv $ λ _, iff.rfl
+
+@[to_additive] instance (H : subgroup G) [encodable H] : encodable H.opposite :=
+encodable.of_equiv H H.opposite_equiv.symm
+
+@[to_additive] lemma left_right_mul {H : subgroup G} (x g : G) (h : H.opposite) :
+  h • (g * x) = g * (h • x) :=
+begin
+  cases h,
+  simp [(•), mul_assoc],
+end
+
+@[to_additive] lemma left_right_mem_preimage {H : subgroup G} (x g : G) (h : H.opposite)
```

Comment:
I think it is nicer to change this lemma into:
```
(λ y, h • y) '' (has_mul.mul g ⁻¹' s) = has_mul.mul g ⁻¹' (λ y, h • y) '' s
```
That should also simplify the place where you use it.
As name maybe `smul_opposite_image_mul_preimage`?

Reply:
I tried changing the lemma to state this but sadly failed to make it work. I added parentheses in a few places but it was still giving error messages...?

Reply:
Thanks!